smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.usertune.element;
+
+import java.net.URI;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+import org.jivesoftware.smackx.usertune.UserTuneManager;
+
+/**
+ * {@link ExtensionElement} that contains the UserTune.
+ * <br>
+ * Instance of UserTuneElement can be created using {@link Builder#build()} method.
+ * To communicate a meaningful UserTune ,{@link #title} and {@link #source} are set as compulsory parameters.
+ */
+public final class UserTuneElement implements ExtensionElement {
+
+    public static final String NAMESPACE = "http://jabber.org/protocol/tune";
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final int length;
+    private final int rating;
+    private final String source;
+    private final String title;
+    private final int track;
+    private final URI uri;
+
+    private UserTuneElement(Builder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isNullUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+
+        XmlStringBuilder userTuneXML = new XmlStringBuilder();
+            if (artist != null) {
+                userTuneXML.element("artist", artist);
+            }
+            if (length  != -1) {
+                userTuneXML.element("length", Integer.toString(length));
+            }
+            if (rating  != -1) {
+                userTuneXML.element("rating",Integer.toString(rating));
+            }
+            userTuneXML.element("source",source);
```

Comment:
What happens, if the user hasn't set a source?
Hint: To solve this, take a look at `XmlStringBuilder.optElement()` ;)

Reply:
Same goes for all the other fields as well.

Reply:
If the user hasn't set source, firstly the `builder.build()` method will give throw an IllegalArguementException with a message that "UserTune should atleast have 'title' and 'source'".
This is because I've considered keeping source as a compulsory element because If the user is listening to a tune, it has to be from somewhere, maybe a website or the local machine. Which means that there will be no case in which the source is undefined. Of course, we can make source optional as well , like we did for other elements. 
Incase you think that there is a way in which source could be undefined , it can be set as an optional element.

Reply:
> This is because I've considered keeping source as a compulsory element because If the user is listening to a tune, it has to be from somewhere, maybe a website or the local machine. Which means that there will be no case in which the source is undefined.

Is this defined in the XEP somewhere? I would be cautious to prevent certain use cases without being able to point at the XEP.

Reply:
No, It is not stated anywhere in the XEP. And I realized that it is not wise to violate it.
Will act on it by keeping only the `title` as a compulsory element and others as optional elements.